src/fat.rs

```diff
@@ -353,50 +357,27 @@ impl<'a> OnDiskDirEntry<'a> {
     }
 
     /// If this is an LFN, get the contents so we can re-assemble the filename.
-    pub fn lfn_contents(&self) -> Option<(bool, u8, [char; 13])> {
-        if self.is_lfn() {
-            let mut buffer = [' '; 13];
-            let is_start = (self.data[0] & 0x40) != 0;
-            let sequence = self.data[0] & 0x1F;
-            // LFNs store UCS-2, so we can map from 16-bit char to 32-bit char without problem.
-            buffer[0] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[1..=2]))).unwrap();
-            buffer[1] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[3..=4]))).unwrap();
-            buffer[2] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[5..=6]))).unwrap();
-            buffer[3] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[7..=8]))).unwrap();
-            buffer[4] = core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[9..=10])))
-                .unwrap();
-            buffer[5] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[14..=15])))
-                    .unwrap();
-            buffer[6] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[16..=17])))
-                    .unwrap();
-            buffer[7] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[18..=19])))
-                    .unwrap();
-            buffer[8] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[20..=21])))
-                    .unwrap();
-            buffer[9] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[22..=23])))
-                    .unwrap();
-            buffer[10] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[24..=25])))
-                    .unwrap();
-            buffer[11] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[28..=29])))
-                    .unwrap();
-            buffer[12] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[30..=31])))
-                    .unwrap();
-            Some((is_start, sequence, buffer))
-        } else {
-            None
+    pub fn lfn_contents(&self) -> Option<(bool, u8, String<U26>)> {
+        if !self.is_lfn() {
+            return None;
         }
+        let is_start = (self.data[0] & 0x40) != 0;
+        let sequence = self.data[0] & 0x1F;
+        let mut buffer = [0u16; 13];
+        let bytes: &mut [u8; 26] = unsafe { core::mem::transmute(&mut buffer) };
```

Comment:
Is this allowed? I thought multiple mut references was UB.

Reply:
I think this also assumes the endianess of the CPU matches the on disk format, which does not always hold.

Reply:
Hmm, where is the other reference ? I think you could get into UB if you used `bytes`, then `buffer` and then `bytes` again, but I don't think this is the case. However, I think that we should at least put these operations inside a block `{}` to avoid future problems, `transmute` can be nasty. Anyways, I might also be missing some thing that would make this UB.

I agree about the endianess problem.

Reply:
Hmm, if `transmute` is tying the lifetime of bytes to the lifetime of the passed reference `&mut buffer` then this is UB. I guess we should avoid `transmute` here, we could probably to the same thing with `as` and raw pointers.

Reply:
Or just use the `from_bytes` and `to_bytes` functions and allow the compiler to optimise it all. I don't think there's a need for the fancy tricks here.

Reply:
> I don't think there's a need for the fancy tricks here.

I agree.